### PR TITLE
stream: add exception policy for midstream flows - v3

### PR DESCRIPTION
**Pull request by jufajardini**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
- exception policy midstream: https://redmine.openinfosecfoundation.org/issues/5468
- applayer parser sections merging: https://redmine.openinfosecfoundation.org/issues/5364
- starts to address documenting the exception policy: https://redmine.openinfosecfoundation.org/issues/5475

Describe changes:
- add a stream.midstream-policy option that allows for controlling if a flow should be fully dropped, passed or bypassed. It is also possible to just ignore the policy (keeps normal behavior) if it is a midstream flow (in IPS mode)
- merge the two existing application layer parser configuration sections in the userguide into one
- add a very brief mention to each of the existing exception policy options (that are configurable via suricata.yaml) to the userguide

Updates from last PR (#7748):
- rework configuration options explanation locations in the suricata.yaml.in
- briefly document those options in the userguide (suricata.yaml documentation page)
- brief update of the defrag section (wasn't listing all options available in the yaml file)
- plus a sneaky commit for merging the applayer sections together (as one of the exception policies applies to them)

suricata-verify-pr: 916
https://github.com/OISF/suricata-verify/pull/916

### CODE REVIEW

**Comment by jufajardini** [comment]
All of the failures are due to the behavior I tried to track with these tests https://github.com/OISF/suricata-verify/pull/915 (and the related ticket). The changes that we added to the exception-policy weren't enough to prevent the logging of applayer events for dropped flows, yet...

Should I remove that check from the SV tests, or start investigating the underlying behavior which may be influencing this?

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpw2_single_stats_chk | | |
| .flow.spare | 1949939 | 1835514 | 106.23% |
| | tlpw2_autofp_stats_chk | | |
| .flow.spare | 1962744 | 1812994 | 108.26% |
| .flow.memuse | 518473728 | 550537728 | 94.18% |
| | tlpr1_stats_chk | | |
| .flow.memuse | 600430528 | 845888448 | 70.98% |
| .http.memuse | 16440 | 49320 | 33.33% |
| | ips_afp_stats_chk | | |
| .flow.end.state.new | 75096 | 10800 | 695.33% |
| .flow.end.tcp_liberal | 300465 | 108000 | 278.21% |
| .flow.memuse | 870597568 | 1081197888 | 80.52% |
| .tcp.reassembly_gap | 958855 | 108000 | 887.83% |
| .app_layer.tx.ftp | 249707 | 263520 | 94.76% |
| .app_layer.tx.dcerpc_tcp | 1009750 | 1101600 | 91.66% |
| | generic_stats_chk | | |
| .capture.kernel_drops | 5451 | 5654519 | 0.1% |
| .flow.end.state.new | 6311 | 14867 | 42.45% |
| .flow.end.tcp_state.syn_sent | 13 | 183 | 7.1% |
| .flow.end.tcp_state.syn_recv | 1 | 52 | 1.92% |
| .flow.end.tcp_state.established | 54018 | 61527 | 87.8% |
| .flow.memuse | 760980288 | 709384128 | 107.27% |
| .tcp.segment_memcap_drop | 0 | 11729 | 0.0% |
| .tcp.reassembly_gap | 82039 | 114099 | 71.9% |
| .tcp.insert_data_normal_fail | 0 | 11358 | 0.0% |
| .app_layer.error.http.parser | 1 | 55 | 1.82% |
| .app_layer.error.smtp.gap | 5 | 61 | 8.2% |
| .app_layer.error.tls.gap | 54240 | 60833 | 89.16% | 

Pipeline 8819
WARNING: THERE IS A KNOWN BAD BASELINE WITH PACKET DROPS. bE MINDFUL OF ANY RESULTS.

File: src/stream-tcp.c [hunk 1]
```diff
@@ -925,7 +926,11 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
     /* SYN/ACK */
     } else if ((p->tcph->th_flags & (TH_SYN|TH_ACK)) == (TH_SYN|TH_ACK)) {
         if (!stream_config.midstream && stream_config.async_oneside == FALSE)
-            return 0;
+            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
```
Comment (victorjulien) [inline src/stream-tcp.c:929]: we still implicitly enable midstream here if the exception policy is set

File: src/stream-tcp.c [hunk 2]
```diff
@@ -1086,7 +1091,11 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
 
     } else if (p->tcph->th_flags & TH_ACK) {
         if (!stream_config.midstream)
-            return 0;
+            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
```
Comment (victorjulien) [inline src/stream-tcp.c:1094]: same here

File: src/util-exception-policy.c [hunk 3]
```diff
@@ -35,6 +36,9 @@ void ExceptionPolicyApply(Packet *p, enum ExceptionPolicy policy, enum PacketDro
                 SCLogDebug("EXCEPTION_POLICY_DROP_FLOW");
                 if (p->flow) {
                     p->flow->flags |= FLOW_ACTION_DROP;
+                    FlowSetNoPayloadInspectionFlag(p->flow);
```
Comment (victorjulien) [inline src/util-exception-policy.c:39]: this should go into its own commit

**Comment by victorjulien** [comment]
I think something like this is needed. It makes all tests pass after updating them for the midstream not being enabled issues.
```diff
diff --git a/src/stream-tcp.c b/src/stream-tcp.c
index 6bde5d8fb1..2dceab7316 100644
--- a/src/stream-tcp.c
+++ b/src/stream-tcp.c
@@ -925,13 +925,19 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
 
     /* SYN/ACK */
     } else if ((p->tcph->th_flags & (TH_SYN|TH_ACK)) == (TH_SYN|TH_ACK)) {
-        if (!stream_config.midstream && stream_config.async_oneside == FALSE)
-            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
-                return 0;
-
         /* Drop reason will only be used if midstream policy is set to fail closed */
         ExceptionPolicyApply(p, stream_config.midstream_policy, PKT_DROP_REASON_STREAM_MIDSTREAM);
 
+        if (!stream_config.midstream && stream_config.async_oneside == FALSE) {
+            SCLogNotice("midstream not enabled, so don't pick up a session");
+            return 0;
+        }
+        if (!(stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_FLOW || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_PACKET)) {
+            SCLogNotice("midstream policy not permissive, so don't pick up a session");
+            return 0;
+        }
+        SCLogNotice("midstream picked up");
+
         if (ssn == NULL) {
             ssn = StreamTcpNewSession(p, stt->ssn_pool_id);
             if (ssn == NULL) {
@@ -1090,13 +1096,19 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
                 ssn->client.last_ack);
 
     } else if (p->tcph->th_flags & TH_ACK) {
-        if (!stream_config.midstream)
-            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
-                return 0;
-
         /* Drop reason will only be used if midstream policy is set to fail closed */
         ExceptionPolicyApply(p, stream_config.midstream_policy, PKT_DROP_REASON_STREAM_MIDSTREAM);
 
+        if (!stream_config.midstream) {
+            SCLogNotice("midstream not enabled, so don't pick up a session");
+            return 0;
+        }
+        if (!(stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_FLOW || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_PACKET)) {
+            SCLogNotice("midstream policy not permissive, so don't pick up a session");
+            return 0;
+        }
+        SCLogNotice("midstream picked up");
+
         if (ssn == NULL) {
             ssn = StreamTcpNewSession(p, stt->ssn_pool_id);
             if (ssn == NULL) {
```

**Comment by victorjulien** [comment]
See also https://github.com/OISF/suricata-verify/pull/916#pullrequestreview-1087656444

**Comment by jufajardini** [comment]
> I think something like this is needed. It makes all tests pass after updating them for the midstream not being enabled issues.
> 
> ```diff
> diff --git a/src/stream-tcp.c b/src/stream-tcp.c
> index 6bde5d8fb1..2dceab7316 100644
> --- a/src/stream-tcp.c
> +++ b/src/stream-tcp.c
> @@ -925,13 +925,19 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
>  
>      /* SYN/ACK */
>      } else if ((p->tcph->th_flags & (TH_SYN|TH_ACK)) == (TH_SYN|TH_ACK)) {
> -        if (!stream_config.midstream && stream_config.async_oneside == FALSE)
> -            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
> -                return 0;
> -
>          /* Drop reason will only be used if midstream policy is set to fail closed */
>          ExceptionPolicyApply(p, stream_config.midstream_policy, PKT_DROP_REASON_STREAM_MIDSTREAM);
>  
> +        if (!stream_config.midstream && stream_config.async_oneside == FALSE) {
> +            SCLogNotice("midstream not enabled, so don't pick up a session");
> +            return 0;
> +        }
> +        if (!(stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_FLOW || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_PACKET)) {
> +            SCLogNotice("midstream policy not permissive, so don't pick up a session");
> +            return 0;
> +        }
> +        SCLogNotice("midstream picked up");
> +
>          if (ssn == NULL) {
>              ssn = StreamTcpNewSession(p, stt->ssn_pool_id);
>              if (ssn == NULL) {
> @@ -1090,13 +1096,19 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
>                  ssn->client.last_ack);
>  
>      } else if (p->tcph->th_flags & TH_ACK) {
> -        if (!stream_config.midstream)
> -            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
> -                return 0;
> -
>          /* Drop reason will only be used if midstream policy is set to fail closed */
>          ExceptionPolicyApply(p, stream_config.midstream_policy, PKT_DROP_REASON_STREAM_MIDSTREAM);
>  
> +        if (!stream_config.midstream) {
> +            SCLogNotice("midstream not enabled, so don't pick up a session");
> +            return 0;
> +        }
> +        if (!(stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_FLOW || stream_config.midstream_policy == EXCEPTION_POLICY_PASS_PACKET)) {
> +            SCLogNotice("midstream policy not permissive, so don't pick up a session");
> +            return 0;
> +        }
> +        SCLogNotice("midstream picked up");
> +
>          if (ssn == NULL) {
>              ssn = StreamTcpNewSession(p, stt->ssn_pool_id);
>              if (ssn == NULL) {
> ```

Thanks, I had totally misunderstood the concepts here 😓

**Comment by jufajardini** [comment]
Replaced by: https://github.com/OISF/suricata/pull/7791
